### PR TITLE
MM-38593: Improve semantics for archived playbooks

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- Replacing delete playbook semantics with archive playbook

#### Notes
- /Not/ changing the db, so we're using `DeleteAt` to signify that a playbook is archived. That seems fine to me because it resembles what we do on the server.
- Also /not/ changing then telemetry, so there is still a delete telemetry event. I'm keeping that because I don't want to cause one more data translation task (and having to remember somewhere
- I'm reusing sing the Badge component, but Badge used to be for runs only. Now it's being used for Runs and Playbooks. I chose to use a union instead of adding 'archived' to the run statuses (which they currently can't be). But since we don't have a concept of playbook "state" I didn't create a playbook state type.

Looks like (compare to the ticket):

image

image

image

image

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-38593

#### Checklist

- [ ] ~~Telemetry updated~~
- [ ] ~~Gated by experimental feature flag~~
- [x] Unit tests updated

### CODE REVIEW

File: server/api/playbook_runs_test.go [hunk 1]
```diff
@@ -761,6 +761,56 @@ func TestPlaybookRuns(t *testing.T) {
 		require.Nil(t, resultPlaybookRun)
 	})
 
+	t.Run("create invalid playbook run - playbook is archived", func(t *testing.T) {
```
Comment (crspeller) [inline server/api/playbook_runs_test.go:764]: Have you tried using the new tests?

File: webapp/src/components/backstage/playbook_runs/status_badge.tsx [hunk 2]
```diff
@@ -9,7 +9,7 @@ import {FormattedMessage} from 'react-intl';
 import {PlaybookRunStatus} from 'src/types/playbook_run';
 
 interface BadgeProps {
-    status: PlaybookRunStatus;
+    status: PlaybookRunStatus | 'Archived';
```
Comment (crspeller) [inline webapp/src/components/backstage/playbook_runs/status_badge.tsx:12]: For ease of use maybe defining a type for this? Call it something like "BadgeType".

**Review by crspeller — APPROVED** [review]

File: webapp/src/components/backstage/playbook_runs/status_badge.tsx (hunk 2 above)
Comment (agarciamontoro) [reply webapp/src/components/backstage/playbook_runs/status_badge.tsx:12]: Agree, and maybe move this file out of the playbook_runs/ directory?

File: webapp/src/components/backstage/playbooks/playbook.tsx [hunk 3]
```diff
@@ -120,9 +121,16 @@ const Playbook = () => {
                             <SubTitle>{subTitle}</SubTitle>
                         </HorizontalBlock>
                     </VerticalBlock>
+                    {
+                        archived &&
+                        <StatusBadge
+                            data-testid='archived-badge'
+                            status='Archived'
+                        />
```
Comment (agarciamontoro) [inline webapp/src/components/backstage/playbooks/playbook.tsx:129]: Doesn't the linter enforce the curly braces here? I thought that was needed

```suggestion
                        <StatusBadge
                            data-testid={'archived-badge'}
                            status={'Archived'}
                        />
```

File: tests-e2e/cypress/integration/channels/rhs_spec.js [hunk 4]
```diff
@@ -180,6 +183,9 @@ describe('channels > rhs', () => {
                 cy.apiFinishRun(playbookRun.id);
             });
 
+            // # Wait because this test is flaky if we move too quickly
+            cy.wait(TIMEOUTS.FIVE_SEC);
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/channels/rhs_spec.js:187]: Wow, do we need five seconds?

File: tests-e2e/cypress/integration/playbooks/overview_spec.js [hunk 5]
```diff
@@ -122,4 +125,48 @@ describe('playbooks > overview', () => {
             cy.findByText('Start run').should('exist');
         });
     });
+
+    describe.only('archiving', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/playbooks/overview_spec.js:129]: Should we remove the only?

File: tests-e2e/cypress/integration/playbooks/overview_spec.js [hunk 6]
```diff
@@ -122,4 +125,48 @@ describe('playbooks > overview', () => {
...
+            // # Programmatically archive it
+            cy.apiArchivePlaybook(testPlaybook.id);
+
+            // # Visit the selected playbook
+            cy.visit(`/playbooks/playbooks/${testPlaybook.id}`);
+
+            // * Verify we're on the right playbook
+            cy.get('[class^="Title-"]').contains(playbookTitle);
+
+            // * Verify we an see the archived badge
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/playbooks/overview_spec.js:153]: Typo
```suggestion
            // * Verify we can see the archived badge
```

File: tests-e2e/cypress/integration/playbooks/overview_spec.js [hunk 7]
```diff
@@ -122,4 +125,48 @@ describe('playbooks > overview', () => {
...
+            cy.findByTestId('edit-playbook').should('be.disabled');
+
+            // # Attempt to edit the playbook
+            cy.apiGetPlaybook(testPlaybook.id).then((playbook) => {
+                // # New title
+                playbook.title = 'new Title!!!';
+
+                // * Verify update fails
+                cy.apiUpdatePlaybook(playbook, 400);
+            });
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/playbooks/overview_spec.js:169]: I really like this, but maybe this should be coded as one of the new server E2E tests?

**Review by agarciamontoro — APPROVED** [review]
Nice! I left a couple of nits and minor suggestions, nothing blocking :) Thank you :tada:

File: server/api/playbook_runs_test.go (hunk 1 above)
Comment (cpoile) [reply server/api/playbook_runs_test.go:764]: Yep, done. Had to fix a pretty bad bug in the client, which shows that using the client in tests like this is useful.

File: webapp/src/components/backstage/playbook_runs/status_badge.tsx (hunk 2 above)
Comment (cpoile) [reply webapp/src/components/backstage/playbook_runs/status_badge.tsx]: Okay, sounds good. We have to convert types now, by using `BadgeType[playbookRun.current_status]` for example, but that's easy.

File: tests-e2e/cypress/integration/channels/rhs_spec.js (hunk 4 above)
Comment (cpoile) [reply tests-e2e/cypress/integration/channels/rhs_spec.js:187]: I'm not sure how much of a delay would make the test consistently pass, but this is safe (and far less time than we're wasting by having to rerun the test)

File: tests-e2e/cypress/integration/playbooks/overview_spec.js (hunk 5 above)
Comment (cpoile) [reply tests-e2e/cypress/integration/playbooks/overview_spec.js]: 🤦

File: tests-e2e/cypress/integration/playbooks/overview_spec.js (hunk 7 above)
Comment (cpoile) [reply tests-e2e/cypress/integration/playbooks/overview_spec.js:169]: yep, done. But I kept this, cause why not. :)

**Review by crspeller — APPROVED** [review]

**Pull request merged**